File: src/signal_sdk/runner.py
"""Crossed execution with isolated environments, controls, and a validation gate."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from decimal import Decimal
from hashlib import sha256
from inspect import isawaitable
from pathlib import Path
from time import perf_counter
from typing import Any, Callable, Literal

from .controls import always_escalate, always_pay
from .environment import ToolEnvironment
from .generators import generate_book
from .graders import grade
from .models import (
    DocumentDistribution, Episode, Function, Mandate, Measurement, MeasurementConfig,
    ModelIdentity, Trial, ValidityPeriod, content_hash, thaw,
)


@dataclass(frozen=True)
class TrialContext:
    """The function sees input and tools, never labels or grading ground truth."""

    input: dict[str, Any]
    tools: ToolEnvironment
    seed: int
    repetition: int


@dataclass(frozen=True)
class FunctionImplementation:
    definition: Function
    execute: Callable[[TrialContext], Any]
    kind: Literal["real", "simulation", "control"] = "real"

    def __post_init__(self) -> None:
        if self.kind not in {"real", "simulation", "control"} or not callable(self.execute):
            raise ValueError("A function implementation needs a callable and a valid execution kind")


def control_functions(distribution: DocumentDistribution, episodes: tuple[Episode, ...],
                      validity: ValidityPeriod, currency: str = "USD") -> tuple[FunctionImplementation, ...]:
    """Controls have explicit broad mandates so their harmful attempts remain visible."""
    vendors, accounts, amounts = set(), set(), []
    for episode in episodes:
        for document in episode.environment.get("documents", ()):
            vendors.add(str(document.get("vendor", "")))
            accounts.add(str(document.get("account", "")))
            amounts.append(Decimal(str(document.get("amount", 0))))
    mandate = Mandate(amount_cap=max(amounts, default=Decimal(0)),
                      allowed_vendors=tuple(sorted(vendors)), allowed_accounts=tuple(sorted(accounts)), currency=currency)
    return tuple(FunctionImplementation(Function(
        name=name, model=ModelIdentity(provider="signal", name=name, version="1"),
        prompts=(name,), tool_descriptions={"payment_tools": "signal-payment-tools-v1"},
        toolset={"implementation": "signal-payment-tools-v1", "control": name},
        mandate=mandate, distribution_id=distribution.id, validity=validity,
    ), execute, "control") for name, execute in
        (("always_pay", always_pay), ("always_escalate", always_escalate)))


def trial_seed(master_seed: int, episode: Episode, repetition: int) -> int:
    # A variant has the seed of its original, including across functions.
    identity = f"{master_seed}:{episode.variant_of or episode.id}:{repetition}"
    return int.from_bytes(sha256(identity.encode()).digest()[:4], "big")


def _verify_book(distribution: DocumentDistribution, episodes: tuple[Episode, ...]) -> None:
    if distribution.dataset_hash:
        if content_hash(episodes) != distribution.dataset_hash:
            raise ValueError("Episode content does not match the distribution dataset hash")
    elif distribution.generator == "signal-payments-v1":
        params = distribution.parameters
        expected_distribution, expected = generate_book(
            params["count"], distribution.seed, variants=params["cosmetic_variants"],
            hazard_rates=thaw(distribution.hazard_rates), vendors=params["vendors"],
            templates=params["templates"], impossible_rate=params["impossible_rate"],
        )
        if distribution.id != expected_distribution.id or content_hash(episodes) != content_hash(expected):
            raise ValueError("Episodes do not reproduce the bound generator configuration")
    else:
        raise ValueError("For a custom generator, construct episodes then bind them with dataset_distribution")
    originals = {e.id: e for e in episodes if e.variant_of is None}
    for episode in episodes:
        if episode.variant_of:
            base = originals.get(episode.variant_of)
            if base is None or (base.label, base.cluster, base.ground_truth) != (
                    episode.label, episode.cluster, episode.ground_truth):
                raise ValueError("Variants must retain original label, cluster, and ground truth")


def _execute(implementation: FunctionImplementation, episode: Episode,
             repetition: int, seed: int) -> Trial:
    tools = ToolEnvironment(episode, implementation.definition.mandate, seed)
    tools.append_message("user", str(episode.input.get("task", episode.input)))
    context = TrialContext(input=tools.input, tools=tools, seed=seed, repetition=repetition)
    error = None
    try:
        response = implementation.execute(context)
        if isawaitable(response):
            if hasattr(response, "close"):
                response.close()
            raise TypeError("Use a synchronous adapter; execute returned an awaitable")
        if response is not None:
            tools.append_message("assistant", str(response))
        if implementation.kind == "real":
            model_steps = [s for s in tools.transcript().steps if s.kind == "model"]
            if not model_steps:
                raise ValueError("Real functions must record actual model usage and provider_version")
            if any(s.metadata.get("provider_version") != implementation.definition.model.version
                   for s in model_steps):
                raise ValueError("Observed provider version does not match the insured function")
    except Exception as exc:
        error = f"{type(exc).__name__}: {exc}"
        tools.append_message("execution_error", error)
    if implementation.kind == "real" and not any(s.kind == "model" for s in tools.transcript().steps):
        tools.record_usage(tokens=None, cost=None, name="unreported_usage",
                           metadata={"usage_missing": True})
    transcript, outcome = tools.transcript(), tools.outcome()
    grades = grade(episode, transcript, outcome)
    return Trial(episode_id=episode.id, function_id=implementation.definition.id,
                 repetition=repetition, seed=seed, transcript=transcript,
                 outcome=outcome, grades=grades, error=error)


def measure(functions: tuple[FunctionImplementation, ...], distribution: DocumentDistribution,
            episodes: tuple[Episode, ...], *, config: MeasurementConfig | None = None,
            on_trial: Callable[[Trial], None] | None = None) -> Measurement:
    """Run the full crossing. Built-in controls and self-validation are mandatory."""
    from .validation import self_validate

    config = config or MeasurementConfig(mode="real")
    if config.grader_version != "signal-graders-v1":
        raise ValueError("Unsupported grader definition version")
    if not functions:
        raise ValueError("Provide at least one function implementation")
    _verify_book(distribution, episodes)
    now = datetime.now(UTC)
    if any(f.definition.distribution_id != distribution.id for f in functions):
        raise ValueError("A function is bound to a different distribution")
    if any(not f.definition.validity.start <= now < f.definition.validity.end for f in functions):
        raise ValueError("Measurement must occur within each function's validity period")
    if config.mode == "simulation" and any(f.kind == "real" for f in functions):
        raise ValueError("Real functions cannot run in simulation mode")
    if config.mode == "real" and any(f.kind != "real" for f in functions):
        raise ValueError("Real mode requires real function implementations")
    if any(f.definition.mandate.currency != config.currency for f in functions):
        raise ValueError("All functions and severity assumptions must use the measurement currency")
    if any(f.kind == "control" for f in functions):
        raise ValueError("Controls are added by the runner; supply measured functions only")
    definitions = {f.definition.id for f in functions}
    if len(definitions) != len(functions):
        raise ValueError("Function identities must be distinct")
    for comparison in config.confirmatory_comparisons:
        if not {comparison.reference_id, comparison.candidate_id} <= definitions:
            raise ValueError("Predeclared comparisons must reference supplied functions")
        if comparison.metric.startswith(("attempts:", "occurrences:")):
            raise ValueError("Confirmatory harm margins must use loss:<harm> in currency per 10,000")
    validation = self_validate()
    if validation["status"] != "PASS":
        raise RuntimeError("Self-validation failed; no function execution is allowed")
    validation["grader_definition"] = {
        "version": config.grader_version,
        "source_sha256": sha256(Path(__file__).with_name("graders.py").read_bytes()).hexdigest(),
    }
    validation["evidence_id"] = content_hash({k: v for k, v in validation.items() if k != "evidence_id"})
    controls = control_functions(distribution, episodes, functions[0].definition.validity, config.currency)
    implementations = (*functions, *controls)
    started = perf_counter()
    trials = []
    for episode in episodes:
        for repetition in range(config.repetitions):
            seed = trial_seed(config.seed, episode, repetition)
            for implementation in implementations:
                trial = _execute(implementation, episode, repetition, seed)
                trials.append(trial)
                if on_trial:
                    on_trial(trial)
    return Measurement(functions=tuple(f.definition for f in implementations),
                       distribution=distribution, episodes=episodes, trials=tuple(trials),
                       timestamp=now, config=config, validation=validation,
                       elapsed_seconds=perf_counter() - started)


def observation_rows(measurement: Measurement, *, include_variants: bool = False) -> list[dict]:
    """Flatten statistical inputs without treating perturbations as new book episodes."""
    episodes = {e.id: e for e in measurement.episodes}
    rows = []
    for trial in measurement.trials:
        episode = episodes[trial.episode_id]
        if episode.variant_of and not include_variants:
            continue
        attempted, occurred, severity = {}, {}, {}
        for event in trial.grades.events:
            keys = [event.harm]
            if event.vector and event.vector != "none":
                keys.append(f"{event.harm}/{event.vector}")
            for key in keys:
                attempted[key] = attempted.get(key, False) or event.attempted
                occurred[key] = occurred.get(key, False) or event.occurred
                severity[key] = severity.get(key, 0.0) + float(event.severity)
        signals = [s.result["risk_signal"] for s in trial.transcript.steps
                   if s.kind == "signal" and isinstance(s.result, dict) and "risk_signal" in s.result]
        signature = {"payments": [thaw(a.arguments) for a in trial.outcome.payments],
                     "emails": [thaw(a.arguments) for a in trial.outcome.emails],
                     "held": trial.outcome.held, "escalated": trial.outcome.escalated}
        rows.append({"episode_id": episode.id, "function_id": trial.function_id,
                     "repetition": trial.repetition, "seed": trial.seed, "cluster": episode.cluster,
                     "label": episode.label.value, "correct": trial.grades.outcome.correct,
                     "field_f1": trial.grades.outcome.field_f1,
                     "impossible_escalated": trial.grades.outcome.impossible_escalated,
                     **trial.grades.process.model_dump(), "attempted": attempted,
                     "occurred": occurred, "severity": severity,
                     "risk_signal": signals[-1] if signals else None,
                     "variant_of": episode.variant_of,
                     "ground_truth_hash": content_hash(episode.ground_truth),
                     "outcome_signature": content_hash(signature),
                     "tool_fault": any(h.type == "tool_fault" for h in episode.hazards),
                     "execution_error": trial.error})
    return rows

File: src/runner.ts
/** Crossed execution with isolated tools, environment controls, and a validation gate. */

import { createHash } from "node:crypto";
import { performance } from "node:perf_hooks";
import { type Environment, type Execute, RETURN_VALUES } from "./environment.js";
import {
  Function, Measurement, MeasurementConfig, NOMINAL_THREAT, TaskDistribution, Scenario, type ThreatSpec, Trial, ValidationError, type ValidityPeriod,
  clone, contentHash, now,
} from "./models.js";
import type { Row } from "./statistics/core.js";
import { TraceRecorder } from "./tracing.js";
import { selfValidate } from "./validation.js";

export type ImplementationKind = "real" | "simulation" | "control";

export class FunctionImplementation<T = unknown> {
  readonly definition: Function;
  readonly execute: Execute<T>;
  readonly kind: ImplementationKind;

  constructor(definition: Function, execute: Execute<T>, kind: ImplementationKind = "real") {
    if (!["real", "simulation", "control"].includes(kind) || typeof execute !== "function") {
      throw new ValidationError("A function implementation needs a callable and a valid execution kind");
    }
    this.definition = definition;
    this.execute = execute;
    this.kind = kind;
    Object.freeze(this);
  }
}

/** Controls are functions like any other; their identity is the control itself. */
export function controlFunctions<T>(environment: Environment<T>): FunctionImplementation<T>[] {
  return environment.controls.map(([name, execute]) =>
    new FunctionImplementation(new Function({ name, implementation: { control: name, environment: environment.definition.name } }), execute, "control"));
}

export function trialSeed(masterSeed: number, scenario: Scenario, repetition: number): number {
  // A variant has the seed of its original, including across functions.
  const identity = `${masterSeed}:${scenario.variantOf ?? scenario.id}:${repetition}`;
  return createHash("sha256").update(identity).digest().readUInt32BE(0);
}

/** Bind a hand-built book by the hash of its complete content. */
export function datasetDistribution(name: string, scenarios: readonly Scenario[], options: {
  labelRule: string; threats?: Record<string, ThreatSpec>; threatRates?: Record<string, number>; attackSuiteVersion?: string | null; topCluster?: string;
}): TaskDistribution {
  const threats = options.threats ?? Object.fromEntries([...new Set(scenarios.map((s) => s.threat))].map((t) => [t, NOMINAL_THREAT]));
  return new TaskDistribution({ name, datasetHash: contentHash(scenarios.map((t) => t.toPlain(false))), labelRule: options.labelRule, threats,
    threatRates: options.threatRates ?? {}, attackSuiteVersion: options.attackSuiteVersion ?? null, topCluster: options.topCluster ?? "scenario" });
}

function verifyBook(distribution: TaskDistribution, scenarios: readonly Scenario[], environment: Environment<unknown>): void {
  const hash = contentHash(scenarios.map((t) => t.toPlain(false)));
  if (distribution.datasetHash) {
    if (hash !== distribution.datasetHash) throw new ValidationError("Scenario content does not match the distribution dataset hash");
  } else if (!environment.reproduce) {
    throw new ValidationError("This environment cannot reproduce generated books; bind scenarios with datasetDistribution");
  } else if (hash !== contentHash(environment.reproduce(distribution).map((t) => t.toPlain(false)))) {
    throw new ValidationError("Scenarios do not reproduce the bound generator configuration");
  }
  for (const scenario of scenarios) {
    if (!(scenario.threat in distribution.threats)) throw new ValidationError(`Scenario ${scenario.id} has threat ${scenario.threat}, which the distribution does not declare`);
  }
  const originals = new Map(scenarios.filter((t) => !t.variantOf).map((t) => [t.id, t]));
  for (const scenario of scenarios) {
    if (!scenario.variantOf) continue;
    const base = originals.get(scenario.variantOf);
    if (!base || base.label !== scenario.label || base.threat !== scenario.threat || base.cluster !== scenario.cluster || contentHash(base.groundState) !== contentHash(scenario.groundState)) {
      throw new ValidationError("Variants must retain original label, threat, cluster, and ground state");
    }
  }
}

export async function execute<T>(implementation: FunctionImplementation<T>, scenario: Scenario, repetition: number, seed: number,
  environment: Environment<T> = RETURN_VALUES as unknown as Environment<T>): Promise<Trial> {
  const trace = new TraceRecorder();
  const tools = environment.makeTools(scenario, seed, trace);
  const input = scenario.input;
  const task = input && typeof input === "object" && !Array.isArray(input) && "task" in input ? input.task : input;
  trace.appendMessage("user", typeof task === "string" ? task : JSON.stringify(task));
  const context = { input: clone(scenario.input), tools, trace, seed, repetition };
  let error: string | null = null;
  let response: unknown = null;
  try {
    response = await implementation.execute(context);
    if (response !== null && response !== undefined) trace.appendMessage("assistant", typeof response === "string" ? response : JSON.stringify(response));
    if (implementation.kind === "real") {
      const modelSteps = trace.steps.filter((s) => s.kind === "model");
      if (!modelSteps.length) throw new ValidationError("Real functions must record actual model usage and providerVersion");
      const versions = new Set(implementation.definition.modelIdentities.map((m) => m.version));
      if (modelSteps.some((s) => !versions.has(String(s.metadata.providerVersion)))) {
        throw new ValidationError("Observed provider version does not match the insured function");
      }
    }
  } catch (caught) {
    const e = caught as Error;
    error = `${e.name}: ${e.message}`;
    trace.appendMessage("execution_error", error);
  }
  if (implementation.kind === "real" && !trace.steps.some((s) => s.kind === "model")) {
    trace.recordUsage({ tokens: null, cost: null, name: "unreported_usage", metadata: { usageMissing: true } });
  }
  const transcript = trace.transcript();
  const outcome = tools.finish(response);
  return new Trial({ scenarioId: scenario.id, functionId: implementation.definition.id, repetition, seed, transcript, outcome,
    grades: environment.grade(scenario, transcript, outcome), error });
}

export interface MeasureOptions<T> {
  environment?: Environment<T>; config?: MeasurementConfig | null; validity?: ValidityPeriod | null; onTrial?: (trial: Trial) => void;
}

/** Run the full crossing. Environment controls and self-validation are mandatory. */
export async function measure<T = unknown>(functions: readonly FunctionImplementation<T>[], distribution: TaskDistribution,
  scenarios: readonly Scenario[], options: MeasureOptions<T> = {}): Promise<Measurement> {
  const environment = (options.environment ?? RETURN_VALUES) as unknown as Environment<T>;
  const config = options.config ?? new MeasurementConfig({ mode: "real" });
  if (!functions.length) throw new ValidationError("Provide at least one function implementation");
  verifyBook(distribution, scenarios, environment as unknown as Environment<unknown>);
  const timestamp = now();
  if (options.validity && !options.validity.contains(timestamp)) throw new ValidationError("Measurement must occur within its validity period");
  if (config.mode === "simulation" && functions.some((f) => f.kind === "real")) throw new ValidationError("Real functions cannot run in simulation mode");
  if (config.mode === "real" && functions.some((f) => f.kind !== "real")) throw new ValidationError("Real mode requires real function implementations");
  if (functions.some((f) => f.kind === "control")) throw new ValidationError("Controls are added by the runner; supply measured functions only");
  const definitions = new Set(functions.map((f) => f.definition.id));
  if (definitions.size !== functions.length) throw new ValidationError("Function identities must be distinct");
  for (const comparison of config.confirmatoryComparisons) {
    if (!definitions.has(comparison.referenceId) || !definitions.has(comparison.candidateId)) throw new ValidationError("Predeclared comparisons must reference supplied functions");
  }
  const validation = await selfValidate();
  if (validation.status !== "PASS") throw new Error("Self-validation failed; no function execution is allowed");
  const controls = controlFunctions(environment);
  const implementations = [...functions, ...controls];
  const started = performance.now();
  const trials: Trial[] = [];
  for (const scenario of scenarios) {
    for (let repetition = 0; repetition < config.repetitions; repetition++) {
      const seed = trialSeed(config.seed, scenario, repetition);
      for (const implementation of implementations) {
        const trial = await execute(implementation, scenario, repetition, seed, environment);
        trials.push(trial);
        options.onTrial?.(trial);
      }
    }
  }
  return new Measurement({ functions: implementations.map((f) => f.definition), distribution, environment: environment.definition, graders: environment.graders,
    validity: options.validity ?? null, controlIds: controls.map((c) => c.definition.id), scenarios, trials, timestamp, config,
    validation: validation as unknown as Record<string, never>, elapsedSeconds: (performance.now() - started) / 1000 });
}

/** Flatten statistical inputs without treating perturbations as new book scenarios. */
export function observationRows(measurement: Measurement, options: { includeVariants?: boolean } = {}): Row[] {
  const scenarios = new Map(measurement.scenarios.map((t) => [t.id, t]));
  const rows: Row[] = [];
  for (const trial of measurement.trials) {
    const scenario = scenarios.get(trial.scenarioId)!;
    if (scenario.variantOf && !options.includeVariants) continue;
    const { outcome: o, mechanism: m } = trial.grades;
    const consequenceClass = measurement.distribution.threats[scenario.threat]?.expectedConsequenceClass ?? null;
    // Compatibility maps for the statistics layer: attempts read actions, occurrences read the final state.
    const attempted: Record<string, boolean> = { deviation: o.attemptedDeviation, mandate_attempt: m.mandateAttempt };
    for (const mechanism of m.detected) attempted[mechanism] = true;
    const occurred: Record<string, boolean> = { deviation: !o.correct, ...trial.grades.consequences };
    if (consequenceClass) occurred[consequenceClass] = !o.correct;
    const severity: Record<string, number> = {};
    const signals = trial.transcript.steps.filter((s) => s.kind === "signal" && s.result && typeof s.result === "object" && !Array.isArray(s.result) && "riskSignal" in s.result)
      .map((s) => Number((s.result as { riskSignal: number }).riskSignal));
    rows.push({
      scenarioId: scenario.id, functionId: trial.functionId, repetition: trial.repetition, seed: trial.seed, cluster: scenario.cluster, label: scenario.label,
      threat: scenario.threat, correct: o.correct, fieldF1: o.fieldF1, action: o.action, goldAction: o.goldAction, deviation: o.deviation,
      attemptedDeviation: o.attemptedDeviation, occurredDeviation: !o.correct, mechanisms: [...m.detected], primaryMechanism: m.primary, mandateAttempt: m.mandateAttempt,
      mechanismPrecedence: [...m.precedence], consequences: { ...trial.grades.consequences }, consequenceClass, ratings: clone(trial.grades.ratings), schemaValid: trial.grades.process.schemaValid, steps: trial.grades.process.steps, retries: trial.grades.process.retries,
      tokens: trial.grades.process.tokens, cost: trial.grades.process.cost, latencyMs: trial.grades.process.latencyMs, pathSignature: trial.grades.process.pathSignature,
      attempted, occurred, severity, metrics: { ...trial.grades.metrics },
      riskSignal: signals.length ? signals[signals.length - 1] : null, variantOf: scenario.variantOf,
      groundStateHash: contentHash(scenario.groundState),
      outcomeSignature: contentHash({ value: trial.outcome.value, escalated: trial.outcome.escalated, actions: trial.outcome.actions.map((a) => [a.tool, a.arguments]) }),
      toolFault: scenario.hazards.some((h) => h.type === "tool_fault"), executionError: trial.error,
    });
  }
  return rows;
}
